File: app/src/main/java/com/krystianwsul/checkme/gui/customtimes/ShowCustomTimeActivity.java
package com.krystianwsul.checkme.gui.customtimes;

import android.content.Context;
import android.content.Intent;
import android.os.Bundle;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.support.design.widget.TextInputLayout;
import android.support.v4.app.LoaderManager;
import android.support.v4.content.Loader;
import android.support.v7.app.ActionBar;
import android.support.v7.widget.Toolbar;
import android.text.Editable;
import android.text.TextUtils;
import android.text.TextWatcher;
import android.view.Menu;
import android.view.MenuItem;
import android.view.View;
import android.widget.EditText;
import android.widget.LinearLayout;
import android.widget.TextView;

import com.krystianwsul.checkme.R;
import com.krystianwsul.checkme.domainmodel.DomainFactory;
import com.krystianwsul.checkme.gui.AbstractActivity;
import com.krystianwsul.checkme.gui.DiscardDialogFragment;
import com.krystianwsul.checkme.gui.TimePickerDialogFragment;
import com.krystianwsul.checkme.loaders.ShowCustomTimeLoader;
import com.krystianwsul.checkme.persistencemodel.SaveService;
import com.krystianwsul.checkme.utils.time.DayOfWeek;
import com.krystianwsul.checkme.utils.time.HourMinute;

import junit.framework.Assert;

import java.util.HashMap;

public class ShowCustomTimeActivity extends AbstractActivity implements LoaderManager.LoaderCallbacks<ShowCustomTimeLoader.Data> {
    private static final String CUSTOM_TIME_ID_KEY = "customTimeId";
    private static final String NEW_KEY = "new";

    private static final String HOUR_MINUTE_SUNDAY_KEY = "hourMinuteSunday";
    private static final String HOUR_MINUTE_MONDAY_KEY = "hourMinuteMonday";
    private static final String HOUR_MINUTE_TUESDAY_KEY = "hourMinuteTuesday";
    private static final String HOUR_MINUTE_WEDNESDAY_KEY = "hourMinuteWednesday";
    private static final String HOUR_MINUTE_THURSDAY_KEY = "hourMinuteThursday";
    private static final String HOUR_MINUTE_FRIDAY_KEY = "hourMinuteFriday";
    private static final String HOUR_MINUTE_SATURDAY_KEY = "hourMinuteSaturday";

    private static final String EDITED_DAY_OF_WEEK_KEY = "editedDayOfWeek";

    private static final String TIME_PICKER_TAG = "timePicker";
    private static final String DISCARD_TAG = "discard";

    public static final int CREATE_CUSTOM_TIME_REQUEST_CODE = 1;

    @Nullable
    private Integer mCustomTimeId;

    @Nullable
    private ShowCustomTimeLoader.Data mData;

    private final HashMap<DayOfWeek, TextView> mTimeViews = new HashMap<>();
    private final HashMap<DayOfWeek, HourMinute> mHourMinutes = new HashMap<>();

    private DayOfWeek mEditedDayOfWeek = null;

    private TextInputLayout mToolbarLayout;
    private EditText mToolbarEditText;

    @Nullable
    private Bundle mSavedInstanceState;

    private static final HourMinute sDefaultHourMinute = new HourMinute(9, 0);

    private final DiscardDialogFragment.DiscardDialogListener mDiscardDialogListener = ShowCustomTimeActivity.this::finish;

    private LinearLayout mShowCustomTimeContainer;

    private final TimePickerDialogFragment.Listener mTimePickerDialogFragmentListener = hourMinute -> {
        Assert.assertTrue(mEditedDayOfWeek != null);
        Assert.assertTrue(mTimeViews.containsKey(mEditedDayOfWeek));
        Assert.assertTrue(mHourMinutes.containsKey(mEditedDayOfWeek));

        mHourMinutes.put(mEditedDayOfWeek, hourMinute);
        mTimeViews.get(mEditedDayOfWeek).setText(hourMinute.toString());

        mEditedDayOfWeek = null;
    };

    @NonNull
    public static Intent getEditIntent(int customTimeId, @NonNull Context context) {
        Intent intent = new Intent(context, ShowCustomTimeActivity.class);
        intent.putExtra(CUSTOM_TIME_ID_KEY, customTimeId);
        return intent;
    }

    @NonNull
    public static Intent getCreateIntent(@NonNull Context context) {
        Intent intent = new Intent(context, ShowCustomTimeActivity.class);
        intent.putExtra(NEW_KEY, true);
        return intent;
    }

    @Override
    public boolean onCreateOptionsMenu(@NonNull Menu menu) {
        getMenuInflater().inflate(R.menu.menu_save, menu);
        return true;
    }

    @Override
    public boolean onPrepareOptionsMenu(@NonNull Menu menu) {
        Assert.assertTrue(mToolbarEditText != null);

        menu.findItem(R.id.action_save).setVisible((mCustomTimeId == null) || (mData != null));

        return true;
    }

    @Override
    public boolean onOptionsItemSelected(@NonNull MenuItem item) {
        switch (item.getItemId()) {
            case R.id.action_save:
                Assert.assertTrue(!mHourMinutes.isEmpty());

                updateError();

                String name = mToolbarEditText.getText().toString().trim();
                if (TextUtils.isEmpty(name))
                    break;

                getSupportLoaderManager().destroyLoader(0);

                if (mData != null) {
                    DomainFactory.getDomainFactory().updateCustomTime(ShowCustomTimeActivity.this, mData.getDataId(), SaveService.Source.GUI, mData.getId(), name, mHourMinutes);
                } else {
                    int customTimeId = DomainFactory.getDomainFactory().createCustomTime(ShowCustomTimeActivity.this, SaveService.Source.GUI, name, mHourMinutes);
                    Assert.assertTrue(customTimeId > 0);

                    setResult(customTimeId);
                }

                finish();
                break;
            case android.R.id.home:
                if (tryClose())
                    finish();
                break;
            default:
                throw new UnsupportedOperationException();
        }
        return true;
    }

    @Override
    protected void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_show_custom_time);

        Toolbar toolbar = findViewById(R.id.toolbar);
        Assert.assertTrue(toolbar != null);

        setSupportActionBar(toolbar);

        ActionBar actionBar = getSupportActionBar();
        Assert.assertTrue(actionBar != null);

        actionBar.setDisplayHomeAsUpEnabled(true);
        actionBar.setHomeAsUpIndicator(R.drawable.ic_close_white_24dp);

        mSavedInstanceState = savedInstanceState;

        mToolbarLayout = findViewById(R.id.toolbarLayout);
        Assert.assertTrue(mToolbarLayout != null);

        mToolbarEditText = findViewById(R.id.toolbarEditText);
        Assert.assertTrue(mToolbarEditText != null);

        mShowCustomTimeContainer = findViewById(R.id.show_custom_time_container);
        Assert.assertTrue(mShowCustomTimeContainer != null);

        initializeDay(DayOfWeek.SUNDAY, R.id.time_sunday_name, R.id.time_sunday_time);
        initializeDay(DayOfWeek.MONDAY, R.id.time_monday_name, R.id.time_monday_time);
        initializeDay(DayOfWeek.TUESDAY, R.id.time_tuesday_name, R.id.time_tuesday_time);
        initializeDay(DayOfWeek.WEDNESDAY, R.id.time_wednesday_name, R.id.time_wednesday_time);
        initializeDay(DayOfWeek.THURSDAY, R.id.time_thursday_name, R.id.time_thursday_time);
        initializeDay(DayOfWeek.FRIDAY, R.id.time_friday_name, R.id.time_friday_time);
        initializeDay(DayOfWeek.SATURDAY, R.id.time_saturday_name, R.id.time_saturday_time);

        Intent intent = getIntent();

        if (mSavedInstanceState != null && mSavedInstanceState.containsKey(HOUR_MINUTE_SUNDAY_KEY)) {
            Assert.assertTrue(mSavedInstanceState.containsKey(EDITED_DAY_OF_WEEK_KEY));
            Assert.assertTrue(mHourMinutes.isEmpty());

            extractKey(HOUR_MINUTE_SUNDAY_KEY, DayOfWeek.SUNDAY);
            extractKey(HOUR_MINUTE_MONDAY_KEY, DayOfWeek.MONDAY);
            extractKey(HOUR_MINUTE_TUESDAY_KEY, DayOfWeek.TUESDAY);
            extractKey(HOUR_MINUTE_WEDNESDAY_KEY, DayOfWeek.WEDNESDAY);
            extractKey(HOUR_MINUTE_THURSDAY_KEY, DayOfWeek.THURSDAY);
            extractKey(HOUR_MINUTE_FRIDAY_KEY, DayOfWeek.FRIDAY);
            extractKey(HOUR_MINUTE_SATURDAY_KEY, DayOfWeek.SATURDAY);

            mEditedDayOfWeek = (DayOfWeek) mSavedInstanceState.getSerializable(EDITED_DAY_OF_WEEK_KEY);

            updateGui();
        } else {
            if (intent.hasExtra(CUSTOM_TIME_ID_KEY)) {
                Assert.assertTrue(!intent.hasExtra(NEW_KEY));
            } else {
                Assert.assertTrue(intent.hasExtra(NEW_KEY));
                Assert.assertTrue(mHourMinutes.isEmpty());

                for (DayOfWeek dayOfWeek : DayOfWeek.values())
                    mHourMinutes.put(dayOfWeek, sDefaultHourMinute);

                updateGui();
            }
        }

        if (intent.hasExtra(CUSTOM_TIME_ID_KEY)) {
            Assert.assertTrue(!intent.hasExtra(NEW_KEY));

            mCustomTimeId = intent.getIntExtra(CUSTOM_TIME_ID_KEY, -1);
            Assert.assertTrue(mCustomTimeId != -1);

            getSupportLoaderManager().initLoader(0, null, this);
        } else {
            Assert.assertTrue(intent.hasExtra(NEW_KEY));
        }

        DiscardDialogFragment discardDialogFragment = (DiscardDialogFragment) getSupportFragmentManager().findFragmentByTag(DISCARD_TAG);
        if (discardDialogFragment != null)
            discardDialogFragment.setDiscardDialogListener(mDiscardDialogListener);
    }

    private void extractKey(@NonNull String key, @NonNull DayOfWeek dayOfWeek) {
        Assert.assertTrue(mSavedInstanceState != null);
        Assert.assertTrue(!TextUtils.isEmpty(key));

        Assert.assertTrue(mSavedInstanceState.containsKey(key));

        HourMinute hourMinute = mSavedInstanceState.getParcelable(key);
        Assert.assertTrue(hourMinute != null);

        mHourMinutes.put(dayOfWeek, hourMinute);
    }

    private void initializeDay(@NonNull final DayOfWeek dayOfWeek, int nameId, int timeId) {
        TextView timeName = findViewById(nameId);
        Assert.assertTrue(timeName != null);

        timeName.setText(dayOfWeek.toString());

        TextView timeView = findViewById(timeId);
        Assert.assertTrue(timeView != null);

        mTimeViews.put(dayOfWeek, timeView);
    }

    @Override
    public void onSaveInstanceState(@NonNull Bundle outState) {
        super.onSaveInstanceState(outState);

        if (!mHourMinutes.isEmpty()) {
            outState.putParcelable(HOUR_MINUTE_SUNDAY_KEY, mHourMinutes.get(DayOfWeek.SUNDAY));
            outState.putParcelable(HOUR_MINUTE_MONDAY_KEY, mHourMinutes.get(DayOfWeek.MONDAY));
            outState.putParcelable(HOUR_MINUTE_TUESDAY_KEY, mHourMinutes.get(DayOfWeek.TUESDAY));
            outState.putParcelable(HOUR_MINUTE_WEDNESDAY_KEY, mHourMinutes.get(DayOfWeek.WEDNESDAY));
            outState.putParcelable(HOUR_MINUTE_THURSDAY_KEY, mHourMinutes.get(DayOfWeek.THURSDAY));
            outState.putParcelable(HOUR_MINUTE_FRIDAY_KEY, mHourMinutes.get(DayOfWeek.FRIDAY));
            outState.putParcelable(HOUR_MINUTE_SATURDAY_KEY, mHourMinutes.get(DayOfWeek.SATURDAY));

            outState.putSerializable(EDITED_DAY_OF_WEEK_KEY, mEditedDayOfWeek);
        }
    }

    @Override
    public Loader<ShowCustomTimeLoader.Data> onCreateLoader(int id, @Nullable Bundle args) {
        return new ShowCustomTimeLoader(this, mCustomTimeId);
    }

    private void updateGui() {
        Assert.assertTrue(!mHourMinutes.isEmpty());

        mToolbarLayout.setVisibility(View.VISIBLE);
        mShowCustomTimeContainer.setVisibility(View.VISIBLE);

        for (DayOfWeek dayOfWeek : DayOfWeek.values()) {
            TextView timeView = mTimeViews.get(dayOfWeek);
            Assert.assertTrue(timeView != null);

            HourMinute hourMinute = mHourMinutes.get(dayOfWeek);
            Assert.assertTrue(hourMinute != null);

            timeView.setText(hourMinute.toString());

            final DayOfWeek finalDayOfWeek = dayOfWeek;
            timeView.setOnClickListener(v -> {
                mEditedDayOfWeek = finalDayOfWeek;

                HourMinute currHourMinute = mHourMinutes.get(finalDayOfWeek);

                TimePickerDialogFragment timePickerDialogFragment = TimePickerDialogFragment.newInstance(currHourMinute);
                timePickerDialogFragment.setListener(mTimePickerDialogFragmentListener);
                timePickerDialogFragment.show(getSupportFragmentManager(), TIME_PICKER_TAG);
            });
        }

        TimePickerDialogFragment timePickerDialogFragment = (TimePickerDialogFragment) getSupportFragmentManager().findFragmentByTag(TIME_PICKER_TAG);
        if (timePickerDialogFragment != null)
            timePickerDialogFragment.setListener(mTimePickerDialogFragmentListener);

        mToolbarEditText.addTextChangedListener(new TextWatcher() {
            private boolean mSkip = (mSavedInstanceState != null);

            @Override
            public void beforeTextChanged(CharSequence s, int start, int count, int after) {

            }

            @Override
            public void onTextChanged(CharSequence s, int start, int before, int count) {

            }

            @Override
            public void afterTextChanged(Editable s) {
                if (mSkip) {
                    mSkip = false;
                    return;
                }

                updateError();
            }
        });
    }

    @Override
    public void onLoadFinished(@NonNull Loader<ShowCustomTimeLoader.Data> loader, @NonNull ShowCustomTimeLoader.Data data) {
        mData = data;

        if (mSavedInstanceState == null || !mSavedInstanceState.containsKey(HOUR_MINUTE_SUNDAY_KEY)) {
            Assert.assertTrue(mHourMinutes.isEmpty());

            mToolbarEditText.setText(mData.getName());

            for (DayOfWeek dayOfWeek : DayOfWeek.values())
                mHourMinutes.put(dayOfWeek, mData.getHourMinutes().get(dayOfWeek));

            updateGui();
        }

        invalidateOptionsMenu();
    }

    @Override
    public void onLoaderReset(@NonNull Loader<ShowCustomTimeLoader.Data> data) {
    }

    @Override
    public void onBackPressed() {
        if (tryClose())
            super.onBackPressed();
    }

    private boolean tryClose() {
        if (dataChanged()) {
            DiscardDialogFragment discardDialogFragment = DiscardDialogFragment.newInstance();
            discardDialogFragment.setDiscardDialogListener(mDiscardDialogListener);
            discardDialogFragment.show(getSupportFragmentManager(), DISCARD_TAG);

            return false;
        } else {
            return true;
        }
    }

    private boolean dataChanged() {
        if (mCustomTimeId == null) {
            Assert.assertTrue(mData == null);

            if (!TextUtils.isEmpty(mToolbarEditText.getText()))
                return true;

            for (DayOfWeek dayOfWeek : DayOfWeek.values())
                if (!mHourMinutes.get(dayOfWeek).equals(sDefaultHourMinute))
                    return true;

            return false;
        } else {
            if (mData == null)
                return false;

            if (!mToolbarEditText.getText().toString().equals(mData.getName()))
                return true;

            for (DayOfWeek dayOfWeek : DayOfWeek.values())
                if (!mHourMinutes.get(dayOfWeek).equals(mData.getHourMinutes().get(dayOfWeek)))
                    return true;

            return false;
        }
    }

    private void updateError() {
        if (TextUtils.isEmpty(mToolbarEditText.getText())) {
            mToolbarLayout.setError(getString(R.string.nameError));
            mToolbarLayout.setPadding(0, 0, 0, 0);
        } else {
            mToolbarLayout.setError(null);
        }
    }
}

File: app/src/main/java/com/krystianwsul/checkme/gui/customtimes/ShowCustomTimeActivity.kt
package com.krystianwsul.checkme.gui.customtimes

import android.app.Activity
import android.content.Context
import android.content.Intent
import android.os.Bundle
import android.os.Parcelable
import android.text.Editable
import android.text.TextWatcher
import android.view.Menu
import android.view.MenuItem
import android.view.View
import androidx.core.view.isVisible
import androidx.transition.TransitionManager
import com.google.android.material.timepicker.MaterialTimePicker
import com.krystianwsul.checkme.R
import com.krystianwsul.checkme.databinding.ActivityShowCustomTimeBinding
import com.krystianwsul.checkme.domainmodel.DomainListenerManager
import com.krystianwsul.checkme.domainmodel.extensions.createCustomTime
import com.krystianwsul.checkme.domainmodel.extensions.updateCustomTime
import com.krystianwsul.checkme.domainmodel.update.AndroidDomainUpdater
import com.krystianwsul.checkme.gui.base.NavBarActivity
import com.krystianwsul.checkme.gui.dialogs.ConfirmDialogFragment
import com.krystianwsul.checkme.gui.dialogs.newMaterialTimePicker
import com.krystianwsul.checkme.gui.dialogs.setListener
import com.krystianwsul.checkme.gui.widgets.MyTextInputLayout
import com.krystianwsul.checkme.utils.tryGetFragment
import com.krystianwsul.checkme.viewmodels.ShowCustomTimeViewModel
import com.krystianwsul.checkme.viewmodels.getViewModel
import com.krystianwsul.common.time.DayOfWeek
import com.krystianwsul.common.time.HourMinute
import com.krystianwsul.common.utils.CustomTimeKey
import io.reactivex.rxjava3.android.schedulers.AndroidSchedulers
import io.reactivex.rxjava3.kotlin.addTo
import io.reactivex.rxjava3.kotlin.plusAssign
import io.reactivex.rxjava3.kotlin.subscribeBy
import java.util.*

class ShowCustomTimeActivity : NavBarActivity() {

    companion object {

        const val CUSTOM_TIME_KEY = "customTimeKey"
        private const val NEW_KEY = "new"

        private const val KEY_HOUR_MINUTES = "hourMinutes"
        private const val KEY_ALL_DAYS_EXPANDED = "allDaysExpanded"

        private const val TAG_TIME_PICKER = "timePicker"
        private const val TAG_TIME_PICKER_ALL_DAYS = "timePickerAllDays"
        private const val DISCARD_TAG = "discard"

        const val CREATE_CUSTOM_TIME_REQUEST_CODE = 1

        private val defaultHourMinute = HourMinute(9, 0)

        fun getEditIntent(
                customTimeKey: CustomTimeKey.Project.Private,
                context: Context,
        ) = Intent(context, ShowCustomTimeActivity::class.java).apply {
            @Suppress("CAST_NEVER_SUCCEEDS")
            putExtra(CUSTOM_TIME_KEY, customTimeKey as Parcelable)
        }

        fun getCreateIntent(context: Context) = Intent(context, ShowCustomTimeActivity::class.java).apply {
            putExtra(NEW_KEY, true)
        }
    }

    private var customTimeKey: CustomTimeKey.Project.Private? = null

    private var data: ShowCustomTimeViewModel.Data? = null

    private lateinit var timeViews: Map<DayOfWeek, MyTextInputLayout>
    private var hourMinutes = mutableMapOf<DayOfWeek, HourMinute>()

    private var savedInstanceState: Bundle? = null

    private val discardDialogListener: (Parcelable?) -> Unit = { finish() }

    private val timePickerDialogFragmentListener = { hourMinute: HourMinute, dayOfWeek: DayOfWeek ->
        check(timeViews.containsKey(dayOfWeek))
        check(hourMinutes.containsKey(dayOfWeek))

        hourMinutes[dayOfWeek] = hourMinute
        timeViews.getValue(dayOfWeek).setText(hourMinute.toString())

        if (dayOfWeek == DayOfWeek.SUNDAY) binding.timeAllDaysTime.setText(hourMinute.toString())
    }

    private val allDaysListener = { hourMinute: HourMinute ->
        DayOfWeek.values().forEach { dayOfWeek ->
            hourMinutes[dayOfWeek] = hourMinute
            timeViews.getValue(dayOfWeek).setText(hourMinute.toString())
        }

        binding.timeAllDaysTime.setText(hourMinute.toString())
    }

    private var showCustomTimeViewModel: ShowCustomTimeViewModel? = null

    private var allDaysExpanded = false

    override val rootView get() = binding.showCustomTimeRoot

    private lateinit var binding: ActivityShowCustomTimeBinding

    override fun onCreateOptionsMenu(menu: Menu): Boolean {
        menuInflater.inflate(R.menu.menu_save, menu)
        return true
    }

    override fun onPrepareOptionsMenu(menu: Menu): Boolean {
        menu.findItem(R.id.action_save).isVisible = customTimeKey == null || data != null

        return true
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        when (item.itemId) {
            R.id.action_save -> {
                check(hourMinutes.isNotEmpty())

                updateError()

                val name = binding.showCustomTimeToolbarEditTextInclude
                        .toolbarEditText
                        .text
                        .toString()
                        .trim { it <= ' ' }

                if (name.isNotEmpty()) {
                    showCustomTimeViewModel?.stop()

                    if (data != null) {
                        AndroidDomainUpdater.updateCustomTime(
                                DomainListenerManager.NotificationType.All,
                                data!!.key,
                                name,
                                hourMinutes
                        )
                                .observeOn(AndroidSchedulers.mainThread())
                                .subscribe { finish() }
                                .addTo(createDisposable)
                    } else {
                        AndroidDomainUpdater.createCustomTime(
                                DomainListenerManager.NotificationType.All,
                                name,
                                hourMinutes,
                        )
                                .observeOn(AndroidSchedulers.mainThread())
                                .subscribeBy {
                                    setResult(
                                            Activity.RESULT_OK,
                                            Intent().putExtra(CUSTOM_TIME_KEY, it as Parcelable),
                                    )

                                    finish()
                                }
                                .addTo(createDisposable)
                    }
                }
            }
            android.R.id.home -> if (tryClose()) finish()
            else -> throw UnsupportedOperationException()
        }
        return true
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)

        binding = ActivityShowCustomTimeBinding.inflate(layoutInflater)
        setContentView(binding.root)

        setSupportActionBar(binding.showCustomTimeToolbarEditTextInclude.toolbar)

        supportActionBar!!.run {
            setDisplayHomeAsUpEnabled(true)
            setHomeAsUpIndicator(R.drawable.ic_close_white_24dp)
        }

        this.savedInstanceState = savedInstanceState

        timeViews = binding.run {
            listOf(
                    Triple(DayOfWeek.SUNDAY, timeSundayName, timeSundayTime),
                    Triple(DayOfWeek.MONDAY, timeMondayName, timeMondayTime),
                    Triple(DayOfWeek.TUESDAY, timeTuesdayName, timeTuesdayTime),
                    Triple(DayOfWeek.WEDNESDAY, timeWednesdayName, timeWednesdayTime),
                    Triple(DayOfWeek.THURSDAY, timeThursdayName, timeThursdayTime),
                    Triple(DayOfWeek.FRIDAY, timeFridayName, timeFridayTime),
                    Triple(DayOfWeek.SATURDAY, timeSaturdayName, timeSaturdayTime)
            )
        }.associate { (dayOfWeek, name, time) ->
            name.text = dayOfWeek.toString()

            dayOfWeek to time
        }

        if (savedInstanceState?.containsKey(KEY_HOUR_MINUTES) == true) {
            check(hourMinutes.isEmpty())

            @Suppress("UNCHECKED_CAST")
            hourMinutes = savedInstanceState.getSerializable(KEY_HOUR_MINUTES) as HashMap<DayOfWeek, HourMinute>

            allDaysExpanded = savedInstanceState.getBoolean(KEY_ALL_DAYS_EXPANDED)

            updateGui()
        } else {
            if (intent.hasExtra(CUSTOM_TIME_KEY)) {
                check(!intent.hasExtra(NEW_KEY))
            } else {
                check(intent.hasExtra(NEW_KEY))
                check(hourMinutes.isEmpty())

                for (dayOfWeek in DayOfWeek.values())
                    hourMinutes[dayOfWeek] = defaultHourMinute

                updateGui()
            }
        }

        if (intent.hasExtra(CUSTOM_TIME_KEY)) {
            check(!intent.hasExtra(NEW_KEY))

            customTimeKey = intent.getParcelableExtra(CUSTOM_TIME_KEY)

            showCustomTimeViewModel = getViewModel<ShowCustomTimeViewModel>().apply {
                start(customTimeKey!!)

                createDisposable += data.subscribe { onLoadFinished(it) }
            }
        } else {
            check(intent.hasExtra(NEW_KEY))

            setUpAllDaysToggle()
        }

        tryGetFragment<ConfirmDialogFragment>(DISCARD_TAG)?.listener = discardDialogListener

        binding.timeAllDaysTimeLayout.setDropdown {
            newMaterialTimePicker(
                    this,
                    supportFragmentManager,
                    TAG_TIME_PICKER_ALL_DAYS,
                    allDaysHourMinute,
            ).setListener(allDaysListener)
        }

        tryGetFragment<MaterialTimePicker>(TAG_TIME_PICKER_ALL_DAYS)?.setListener(allDaysListener)

        DayOfWeek.values().forEach { dayOfWeek ->
            timeViews.getValue(dayOfWeek).setDropdown {
                newMaterialTimePicker(
                        this,
                        supportFragmentManager,
                        TAG_TIME_PICKER,
                        hourMinutes.getValue(dayOfWeek),
                        dayOfWeek,
                ).setListener(timePickerDialogFragmentListener)
            }
        }

        tryGetFragment<MaterialTimePicker>(TAG_TIME_PICKER)?.setListener(timePickerDialogFragmentListener)

        binding.showCustomTimeToolbarEditTextInclude
                .toolbarEditText
                .addTextChangedListener(object : TextWatcher {
                    private var skip = savedInstanceState != null

                    override fun beforeTextChanged(s: CharSequence, start: Int, count: Int, after: Int) = Unit

                    override fun onTextChanged(s: CharSequence, start: Int, before: Int, count: Int) = Unit

                    override fun afterTextChanged(s: Editable) {
                        if (skip) skip = false else updateError()
                    }
                })
    }

    public override fun onSaveInstanceState(outState: Bundle) {
        super.onSaveInstanceState(outState)

        if (hourMinutes.isNotEmpty()) {
            outState.putSerializable(KEY_HOUR_MINUTES, HashMap(hourMinutes))
            outState.putBoolean(KEY_ALL_DAYS_EXPANDED, allDaysExpanded)
        }
    }

    private val allDaysHourMinute get() = hourMinutes.getValue(DayOfWeek.SUNDAY)

    private fun updateGui() {
        check(hourMinutes.isNotEmpty())

        binding.showCustomTimeToolbarEditTextInclude
                .toolbarLayout
                .visibility = View.VISIBLE

        binding.showCustomTimeContainer.visibility = View.VISIBLE

        TransitionManager.beginDelayedTransition(binding.showCustomTimeContainer)
        binding.timeIndividualDaysLayout.isVisible = allDaysExpanded
        binding.timeAllDaysTimeLayout.isVisible = !allDaysExpanded

        binding.timeAllDaysTime.setText(allDaysHourMinute.toString())

        for (dayOfWeek in DayOfWeek.values()) {
            val timeView = timeViews.getValue(dayOfWeek)
            val hourMinute = hourMinutes.getValue(dayOfWeek)

            timeView.setText(hourMinute.toString())
        }
    }

    private fun setUpAllDaysToggle() {
        if (allDaysExpanded) binding.timeAllDaysTextLayout.setChecked(true)

        binding.timeAllDaysTextLayout.setDropdown {
            allDaysExpanded = !allDaysExpanded
            binding.timeAllDaysTextLayout.setChecked(allDaysExpanded)

            if (!allDaysExpanded) {
                val hourMinute = hourMinutes.getValue(DayOfWeek.SUNDAY)

                hourMinutes = DayOfWeek.values()
                        .associate { it to hourMinute }
                        .toMutableMap()
            }

            updateGui()
        }
    }

    private fun onLoadFinished(data: ShowCustomTimeViewModel.Data) {
        this.data = data

        if (savedInstanceState?.containsKey(KEY_HOUR_MINUTES) != true) {
            check(hourMinutes.isEmpty())

            binding.showCustomTimeToolbarEditTextInclude
                    .toolbarEditText
                    .setText(data.name)

            hourMinutes = HashMap(data.hourMinutes)

            allDaysExpanded = hourMinutes.values
                    .distinct()
                    .size > 1

            updateGui()

            setUpAllDaysToggle()
        }

        invalidateOptionsMenu()
    }

    override fun onBackPressed() {
        if (tryClose()) super.onBackPressed()
    }

    private fun tryClose() = if (dataChanged()) {
        ConfirmDialogFragment.newInstance(ConfirmDialogFragment.Parameters.Discard).also {
            it.listener = discardDialogListener
            it.show(supportFragmentManager, DISCARD_TAG)
        }

        false
    } else {
        true
    }

    private fun dataChanged(): Boolean {
        val name = binding.showCustomTimeToolbarEditTextInclude
                .toolbarEditText
                .text

        if (customTimeKey == null) {
            check(data == null)

            if (!name.isNullOrEmpty()) return true

            return DayOfWeek.values().any { hourMinutes[it] != defaultHourMinute }
        } else {
            if (data == null) return false

            if (name.toString() != data!!.name) return true

            return DayOfWeek.values().any { hourMinutes[it] != data!!.hourMinutes[it] }
        }
    }

    private fun updateError() {
        binding.showCustomTimeToolbarEditTextInclude
                .toolbarLayout
                .apply {
                    if (binding.showCustomTimeToolbarEditTextInclude.toolbarEditText.text.isNullOrEmpty()) {
                        error = getString(R.string.nameError)
                        setPadding(0, 0, 0, 0)
                    } else {
                        error = null
                    }
                }
    }
}